File: kill_Down_with_Trojans.py
import numpy as np
import scipy
import sys

#np.set_printoptions(threshold=sys.maxsize)


def load_input_file(file_name):
    with open(file_name, 'r') as file:
        n, H = map(int, file.readline().split())
        tile_types = np.zeros((n, n), dtype=int)
        tile_values = np.zeros((n, n), dtype=int)

        for i in range(n * n):
            if i == 0:
                continue  # the initial tile is zero type with zero value
            x, y, t, v = map(int, file.readline().split())
            tile_types[x][y] = t
            tile_values[x][y] = v

    return n, H, tile_types, tile_values


def print_tile_data(tile_types, tile_values):
    print("Tile Types:")
    print(tile_types)
    print("\nTile Values:")
    print(tile_values)


def DP(n, H, tile_types, tile_values):
    #if player initially has neg health, return False
    if H < 0:
        return False
    memo = np.full((n, n, 2, 2), np.nan)
    result = DP_helper(n, H, tile_types, tile_values, 0, 0, 0, 0, memo)
    if result > H or np.isnan(memo[0][0][0][0]):
        return False
    
    return True

def DP_helper(n, H, tile_types, tile_values, i, j, prev_protect, prev_mult, memo):
    # base cases
    # if reached bottom right corner, return 0
    if (i == n and j == n-1) or (i == n-1 and j == n):
        return 0
    # if out of bounds
    if i == n or j == n:
        return np.inf
    #check memo
    if not np.isnan(memo[i][j][prev_protect][prev_mult]):
        return memo[i][j][prev_protect][prev_mult]
    
    res1 = np.inf
    res2 = np.inf
    res3 = np.inf
    res4 = np.inf

    #if landed on protection square
    if tile_types[i][j] == 2:
        #protect_status = True
        #test going right
        res1 = DP_helper(n, H, tile_types, tile_values, i, j+1, 1, prev_mult, memo)
        #test going down
        res2 = DP_helper(n, H, tile_types, tile_values, i+1, j, 1, prev_mult, memo)
    #if landed on multiplier square
    elif tile_types[i][j] == 3:
        #test going right
        res1 = DP_helper(n, H, tile_types, tile_values, i, j+1, prev_protect, 1, memo)
        #test going down
        res2 = DP_helper(n, H, tile_types, tile_values, i+1, j, prev_protect, 1, memo)
     #if landed on damage tile
    elif tile_types[i][j] == 0:
        #if protected
        if prev_protect:
            #test going right and protect on current tile
            res1 = DP_helper(n, H, tile_types, tile_values, i, j+1, 0, prev_mult, memo)
            #test going down and protect on current tile
            res2 = DP_helper(n, H, tile_types, tile_values, i+1, j, 0, prev_mult, memo)
            #test going right and dont use protect
            res3 = DP_helper(n, H - tile_values[i][j], tile_types, tile_values, i, j+1, prev_protect, prev_mult, memo) + tile_values[i][j]
            #test going down and dont use protect
            res4 = DP_helper(n, H - tile_values[i][j], tile_types, tile_values, i+1, j, prev_protect, prev_mult, memo) + tile_values[i][j]
        else:
            #test going right and not protecting
            res3 = DP_helper(n, H - tile_values[i][j], tile_types, tile_values, i, j+1, prev_protect, prev_mult, memo) + tile_values[i][j]
            #test going down and not protecting
            res4 = DP_helper(n, H - tile_values[i][j], tile_types, tile_values, i+1, j, prev_protect, prev_mult, memo) + tile_values[i][j]
    #if landed on healing tile
    elif tile_types[i][j] == 1:
        #if have multiplier token
        if prev_mult:
            #test going right and multiplier on current tile
            res1 = DP_helper(n, H + tile_values[i][j]*2, tile_types, tile_values, i, j+1, prev_protect, 0, memo) - tile_values[i][j]*2
            #test going down and multiplier on current tile
            res2 = DP_helper(n, H + tile_values[i][j]*2, tile_types, tile_values, i+1, j, prev_protect, 0, memo) - tile_values[i][j]*2
            #test going right and dont use multiplier
            res3 = DP_helper(n, H + tile_values[i][j], tile_types, tile_values, i, j+1, prev_protect, prev_mult, memo) - tile_values[i][j]
            #test going down and dont use multiplier
            res4 = DP_helper(n, H + tile_values[i][j], tile_types, tile_values, i+1, j, prev_protect, prev_mult, memo) - tile_values[i][j]        
        else:
            #test going right and not use multiplier
            res3 = DP_helper(n, H + tile_values[i][j], tile_types, tile_values, i, j+1, prev_protect, prev_mult, memo) - tile_values[i][j]
            #test going down and not use multiplier
            res4 = DP_helper(n, H + tile_values[i][j], tile_types, tile_values, i+1, j, prev_protect, prev_mult, memo) - tile_values[i][j]

    memo[i][j][prev_protect][prev_mult] = max(min(res1,res2,res3,res4),0)
    return memo[i][j][prev_protect][prev_mult]
    


def write_output_file(output_file_name, result):
    with open(output_file_name, 'w') as file:
        file.write(str(int(result)))


def main(input_file_name):
    n, H, tile_types, tile_values = load_input_file(input_file_name)
    print_tile_data(tile_types, tile_values)
    result = DP(n, H, tile_types, tile_values)
    print("Result: " + str(result))
    output_file_name = input_file_name.replace(".txt", "_out.txt")
    write_output_file(output_file_name, result)


if __name__ == "__main__":
    import sys
    if len(sys.argv) != 2:
        print("Usage: python kill_Down_with_Trojans.py a_file_name.txt")
    else:
        main(sys.argv[1])

File: kill_Down_with_Trojans.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cstring> // For strlen

constexpr int MAX_N = 100;

struct Tile
{
    int t;
    int v;
};

void load_input_file(const std::string &file_name, int &n, int &H, std::vector<std::vector<Tile>> &tiles)
{
    std::ifstream file(file_name);
    file >> n >> H;

    // Initialize the grid
    tiles.resize(n, std::vector<Tile>(n));

    for (int i = 0; i < n * n; i++)
    {
        if (i == 0)
        {
            continue; // The initial tile is zero type with zero value
        }
        Tile tile;
        int x, y;
        file >> x >> y >> tile.t >> tile.v;
        tiles[x][y] = tile;
    }

    file.close();
}

void print_tile_data(int n, const std::vector<std::vector<Tile>> &tiles)
{
    std::cout << "Tile Types:\n";
    for (int x = 0; x < n; x++)
    {
        for (int y = 0; y < n; y++)
        {
            std::cout << "[" << tiles[x][y].t << "] ";
        }
        std::cout << '\n';
    }

    std::cout << "\nTile Values:\n";
    for (int x = 0; x < n; x++)
    {
        for (int y = 0; y < n; y++)
        {
            std::cout << "[" << tiles[x][y].v << "] ";
        }
        std::cout << '\n';
    }
}
void printMemo(const std::vector<std::vector<std::vector<std::vector<int>>>> &memo)
{
    for (int i = 0; i < memo.size(); ++i)
    {
        for (int j = 0; j < memo[i].size(); ++j)
        {
            for (int p = 0; p < memo[i][j].size(); ++p)
            {
                for (int m = 0; m < memo[i][j][p].size(); ++m)
                {
                    std::cout << "memo[" << i << "][" << j << "][" << p << "][" << m << "] = " << memo[i][j][p][m] << " ";
                }
                std::cout << std::endl;
            }
            std::cout << std::endl;
        }
        std::cout << std::endl;
    }
}
int helperFunction(int i, int j, int n, bool p, bool m, const std::vector<std::vector<Tile>> &tiles, std::vector<std::vector<std::vector<std::vector<int>>>> &memo)
{

    // Out of bounds
    if (i < 0 || j < 0 || i >= n || j >= n)
    {
        return INT_MAX;
    }

    // If this state is already calculated
    if (memo[i][j][p][m] != -1)
    {
        return memo[i][j][p][m];
    }

    // Base Case: if you've reached the end
    if (i == n - 1 && j == n - 1)
    {
        if (tiles[i][j].t == 0)
        {
            return memo[i][j][p][m] = (p ? 0 : tiles[i][j].v);
        }
        return 0;
    }

    int cost = INT_MAX;

    switch (tiles[i][j].t)
    {
    case 0:
        if (p)
        {
            int useTokenNow = std::min(
                helperFunction(i + 1, j, n, false, m, tiles, memo),
                helperFunction(i, j + 1, n, false, m, tiles, memo));

            // Save the p token for future and take the damage for the current tile
            int saveToken = tiles[i][j].v + std::min(
                                                helperFunction(i + 1, j, n, p, m, tiles, memo),
                                                helperFunction(i, j + 1, n, p, m, tiles, memo));

            cost = std::min(useTokenNow, saveToken);
        }
        else
        {
            cost = tiles[i][j].v + std::min(
                                       helperFunction(i + 1, j, n, p, m, tiles, memo),
                                       helperFunction(i, j + 1, n, p, m, tiles, memo));
        }
        break;

    case 1:
        if (m)
        {
            int useTokenNow = std::min(
                                  helperFunction(i + 1, j, n, p, false, tiles, memo),
                                  helperFunction(i, j + 1, n, p, false, tiles, memo)) -
                              2 * tiles[i][j].v;

            // Save the p token for future and take the damage for the current tile
            int saveToken = std::min(
                                helperFunction(i + 1, j, n, p, m, tiles, memo),
                                helperFunction(i, j + 1, n, p, m, tiles, memo)) -
                            tiles[i][j].v;

            cost = std::min(useTokenNow, saveToken);
        }
        else
        {
            cost = std::min(
                       helperFunction(i + 1, j, n, p, m, tiles, memo),
                       helperFunction(i, j + 1, n, p, m, tiles, memo)) -
                   tiles[i][j].v;
        }
        break;

    case 2:
        cost = std::min(
            helperFunction(i + 1, j, n, true, m, tiles, memo),
            helperFunction(i, j + 1, n, true, m, tiles, memo));
        break;

    case 3:
        cost = std::min(
            helperFunction(i + 1, j, n, p, true, tiles, memo),
            helperFunction(i, j + 1, n, p, true, tiles, memo));
        break;
    }

    cost = std::max(cost, 0);

    return memo[i][j][p][m] = cost;
}
bool DP(int n, int H, const std::vector<std::vector<Tile>> &tiles)
{
    std::vector<std::vector<std::vector<std::vector<int>>>> memo(n, std::vector<std::vector<std::vector<int>>>(n, std::vector<std::vector<int>>(2, std::vector<int>(2, -1))));
    int result = helperFunction(0, 0, n, 0, 0, tiles, memo);
    std::cout << "answer: " << result << '\n';
    // printMemo(memo);
    return (H - result) >= 0;
}

void write_output_file(const std::string &output_file_name, bool result)
{
    std::ofstream file(output_file_name);
    file << result;
    file.close();
}

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        std::cout << "Usage: " << argv[0] << " a_file_name.txt\n";
        return 1;
    }

    int n, H;
    std::vector<std::vector<Tile>> tiles;
    load_input_file(argv[1], n, H, tiles);
    print_tile_data(n, tiles);
    bool result = DP(n, H, tiles);
    std::cout << "Result: " << result << '\n';

    // Construct the output file name
    std::string output_file_name(argv[1]);
    output_file_name.insert(output_file_name.find_last_of('.'), "_out");
    std::ofstream outfile(output_file_name);
    outfile << result;
    outfile.close();

    return 0;
}
